[chore] add unit tests for the opampsupervisor

bacherfl opened the pull request:
**Description:** This PR adds more unit tests to the OpAmpSupervisor. For increasing the coverage further, and to simplify testing in general it might be worth looking into decoupling the supervisor more by introducing interfaces for e.g. the `Commander` and making use of a mocking framework (e.g. https://github.com/uber-go/mock which is already used in other modules within this repo).

**Link to tracking Issue:** #24295 

**Testing:** Added more unit tests to increase coverage

bacherfl committed:
implement unit tests

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
add server tests

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
add license header

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
fix linting

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
fix linting and add changelog entry

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

atoulme commented on .chloggen/opamp-supervisor-unit-tests.yaml:
```diff
@@ -0,0 +1,27 @@
+component: opemapsupervisor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Increase unit test coverage.
```
no need for a changelog for tests, we can skip changelog for those.

bacherfl committed:
remove changelog entry

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
trigger CI checks

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

# Conflicts:
#	cmd/opampsupervisor/supervisor/supervisor_test.go

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

BinaryFissionGames reviewed (changes requested):
Thanks for taking this on, sorry for the delay in review! It generally looks good to me, just a few small things.

bacherfl committed:
fix linting and use staticPIDProvider

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
fix linting

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
trigger CI checks

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl commented:
> Thanks for taking this on, sorry for the delay in review! It generally looks good to me, just a few small things.

No worries, and thanks a lot for the review! I have addressed all comments now (will keep an eye on the CI checks to make sure everything passes)

bacherfl commented on cmd/opampsupervisor/supervisor/supervisor_test.go:
```diff
@@ -291,6 +296,281 @@ func Test_onMessage(t *testing.T) {
+
+		s := Supervisor{
+			logger:      zap.NewNop(),
+			pidProvider: defaultPIDProvider{},
```
good idea, seems like I overlooked the `staticPIDProvider` - that definitely makes it a bit simpler

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

# Conflicts:
#	cmd/opampsupervisor/supervisor/supervisor_test.go

bacherfl committed:
apply suggestions from code review

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl commented on cmd/opampsupervisor/supervisor/supervisor_test.go:
```diff
@@ -386,6 +660,242 @@ func Test_handleAgentOpAMPMessage(t *testing.T) {
+		assert.Equal(t, "test", s.effectiveConfig.Load())
+		assert.True(t, updatedClientEffectiveConfig)
+	})
+	t.Run("EffectiveConfig - Effective config message does not contain instance config", func(t *testing.T) {
```
thanks for the suggestion, that sounds better

bacherfl commented on cmd/opampsupervisor/supervisor/supervisor_test.go:
```diff
@@ -291,6 +296,275 @@ func Test_onMessage(t *testing.T) {
+		fileContent, err := os.ReadFile(filepath.Join(configStorageDir, lastRecvRemoteConfigFile))
+		require.NoError(t, err)
+		assert.Contains(t, string(fileContent), testConfigMessage)
+		// replace the ppid as that changes with each run
```
that comment was a leftover from before where i did not use the static PPID provider yet. I removed that comment now as it's not required anymore

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

# Conflicts:
#	cmd/opampsupervisor/supervisor/supervisor_test.go

bacherfl committed:
adapt to changes from main branch

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

evan-bradley committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

bacherfl committed:
Merge branch 'main' into test/24295/opampsupervisor-tests

bacherfl committed:
adapt failing unit tests

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>
